### PR TITLE
[AArch64][LoopIdiom] Generalize AArch64LoopIdiomTransform into LoopIdiomVectorize

### PR DESCRIPTION
**Pull request by mshockwave**
To facilitate sharing LoopIdiomTransform between AArch64 and RISC-V, this first patch moves AArch64LoopIdiomTransform from lib/Target/AArch64 to lib/Transforms/Vectorize and renames it to LoopIdiomVectorize. The next patch (#94082) will teach LoopIdiomVectorize how to generate VP intrinsics (in addition to the current masked vector style) in favor of RVV. The key component that dictates the vectorization style, of which SVE and RVV differ, is factored out in this patch as well.

### CODE REVIEW

**Review by paperchalice — APPROVED** [review]

**Comment by nikic** [comment]
Is it possible to make the pass name more meaningful? Otherwise we end up with two passes called LoopIdiomRecognize and LoopIdiomTransform (loop-idiom and loop-idiom-transform), and it's not at all obvious how they differ. If I understand correctly, the relevant difference is that the latter is targeted at vector idioms?

(Alternatively, if this is no longer AArch64 specific, can it be merged into the general LoopIdiom pass? But that's probably a more involved change.)

**Comment by preames** [comment]
Can we please split this into two changes, one which does the move and rename, and a second which adds the new functionality?  This diff is slightly confusing as a reader.

Naming wise, maybe LoopIdiomVectorizer?  And update the comment to be really explicit about the goal being to eventually kill this off as the vectorizer grows more powerful?

**Comment by mshockwave** [comment]
> Can we please split this into two changes, one which does the move and rename, and a second which adds the new functionality? This diff is slightly confusing as a reader.

I don't think I add any new functionalities in this patch, the support for RVV-style vectorization is in #94082 

> 
> Naming wise, maybe LoopIdiomVectorizer? And update the comment to be really explicit about the goal being to eventually kill this off as the vectorizer grows more powerful?

I second for LoopIdiomVectorizer.

**Comment by preames** [comment]
> > Can we please split this into two changes, one which does the move and rename, and a second which adds the new functionality? This diff is slightly confusing as a reader.
> 
> I don't think I add any new functionalities in this patch, the support for RVV-style vectorization is in #94082 . Or are you referring to factoring out the part that will differ between SVE and RVV? That change is NFC.

Apologies, it looks like I misread the patch description.

**Comment by mshockwave** [comment]
> (Alternatively, if this is no longer AArch64 specific, can it be merged into the general LoopIdiom pass? But that's probably a more involved change.)

Sorry I missed this question: I tried this but doesn't feel like what LoopIdiom is doing now (pattern matching and turn into libcalls) is somewhat related but not directed related to what this Pass does (pattern matching and turn into vector instructions). And yes, it's a more involved change. So I'm more inclined to not do the merge, at least for now.

**Comment by mshockwave** [comment]
I've renamed the Pass to LoopIdiomVectorize and add a statement in the file header saying that our ultimately goal is to do this in LV instead.

File: llvm/lib/Transforms/Vectorize/LoopIdiomVectorize.cpp [hunk 1]
```diff
@@ -742,6 +682,11 @@ void AArch64LoopIdiomTransform::transformByteCompare(
   DomTreeUpdater DTU(DT, DomTreeUpdater::UpdateStrategy::Lazy);
   Builder.SetCurrentDebugLocation(PHBranch->getDebugLoc());
 
+  // Safeguard to check if we build the correct DomTree with DTU.
+  auto CheckDTU = llvm::make_scope_exit([&]() {
```
Comment (topperc) [inline llvm/lib/Transforms/Vectorize/LoopIdiomVectorize.cpp:686]: Should this be a separate patch?

Comment (mshockwave) [reply llvm/lib/Transforms/Vectorize/LoopIdiomVectorize.cpp]: I agree. These DTU tree changes are now reverted.

**Comment by mshockwave** [comment]
I have now reverted all the intrusive refactoring. I'm planning to create two additional PRs:
  1. Factoring out the part that differs between masked v.s. VP vectorization. Will be NFC.
  2. I found that some of the DominatorTreeUpdates are redundant. I will put such simplifications in a separate PR. But it's still effectively a NFC.

**Review by preames — APPROVED** [review]
LGTM

**Comment by mshockwave** [comment]
I'll wait for another day or so to see if there are any other comments, as I saw a reviewer from ARM has being tagged.

> I have now reverted all the intrusive refactoring. I'm planning to create two additional PRs:
> 
> 1. Factoring out the part that differs between masked v.s. VP vectorization. Will be NFC.

This is now in #94682 

> 2. I found that some of the DominatorTreeUpdates are redundant. I will put such simplifications in a separate PR. But it's still effectively a NFC.

This is now in #94681
